contrib/fprintd/template.py

```diff
@@ -0,0 +1,23 @@
+pkgname = "fprintd"
+pkgver = "1.94.2"
+pkgrel = 0
+build_style = "meson"
+configure_args = ["-Dsystemd=false", "-Dpam_modules_dir=/usr/lib/security"]
+hostmakedepends = ["bash", "gettext", "meson", "perl", "pkgconf"]
+makedepends = [
+    "dbus-devel",
+    "elogind-devel",
+    "glib-devel",
+    "libfprint-devel",
+    "linux-pam-devel",
+    "polkit-devel",
+]
+depends = ["dbus"]
+pkgdesc = "D-Bus service to access fingerprint readers"
+maintainer = "triallax <triallax@tutanota.com>"
+license = "GPL-2.0-or-later"
+url = "https://fprint.freedesktop.org"
+source = f"https://gitlab.freedesktop.org/libfprint/fprintd/-/archive/v{pkgver}/fprintd-v{pkgver}.tar.bz2"
+sha256 = "e388694c60322359c632386fa71c8e1e2cf2abe34363028f1046b0203d3fc640"
+# TODO
```

Comment:
what is the TODO?

Reply:
Many of the tests were mysteriously failing with an error that I couldn't figure out how to fix, but I forgot what it was. I can take another look at it if you want, or otherwise change the comment to something more specific.

Reply:
i'm curious about the error, just in case

Reply:
Here's a sample error (pretty much all of them look like this):
```
370/530 fprintd:daemon+fprintd+FPrintdVirtualDeviceStorageTest / FPrintdVirtualDeviceStorageTest.test_unallowed_verify_with_enroll_claim                                                       FAIL            0.81s   exit status 1
>>> G_SLICE=always-malloc G_DEBUG=fatal-criticals UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 G_MESSAGES_DEBUG=all MALLOC_PERTURB_=170 ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 MALLOC_CHECK_=2 TOPSRCDIR=/
builddir/fprintd-1.94.2 FPRINT_BUILD_DIR=/builddir/fprintd-1.94.2/build/src /usr/bin/python3 /builddir/fprintd-1.94.2/tests/fprintd.py FPrintdVirtualDeviceStorageTest.test_unallowed_verify_with_enroll_claim                        
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――                                                                                                         
stdout:                                                                                                                            
Testing local build                                                                                                                                                   
Using template from /builddir/fprintd-1.94.2/tests/dbusmock/polkitd.py                                                                                   
(fprintd:5769): GLib-GIO-DEBUG: 11:31:48.561: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)                                                                                                                          
(fprintd:5769): fprintd-DEBUG: 11:31:48.562: About to load configuration file '/etc/fprintd.conf'                                                                                                                                    
                                                                                                                                                                                       
(fprintd:5769): fprintd-WARNING **: 11:31:48.562: Could not open "/etc/fprintd.conf": No such file or directory                              
                                                                                                                                   
(fprintd:5769): fprintd-DEBUG: 11:31:48.562: Launching FprintObject                                                                                                   
(fprintd:5769): libfprint-context-DEBUG: 11:31:48.562: Initializing FpContext (libfprint version 1.94.6)                                                 
(fprintd:5769): fprintd-DEBUG: 11:31:48.562: Preparing devices for resume                                                          
1706527908.563 Inhibit "sleep" "net.reactivated.Fprint" "Suspend fingerprint readers" "delay"                                                                                                                                        
(fprintd:5769): libfprint-context-DEBUG: 11:31:48.562: Found virtual environment device: FP_VIRTUAL_DEVICE_STORAGE, /tmp/libfprint-1sn2imk0/virtual-image.socket                                                                                                      
(fprintd:5769): libfprint-context-DEBUG: 11:31:48.563: created                                                                                                        
(fprintd:5769): libfprint-device-DEBUG: 11:31:48.563: Device reported probe completion                                                                                                 
(fprintd:5769): libfprint-device-DEBUG: 11:31:48.563: Completing action FPI_DEVICE_ACTION_PROBE in idle!                                                              
(fprintd:5769): libfprint-device-DEBUG: 11:31:48.563: Updated temperature model after 0.00 seconds, ratio 0.27 -> 0.27, active 0 -> 0, FP_TEMPERATURE_COLD -> FP_TEMPERATURE_COLD                                                                                     
(fprintd:5769): fprintd-DEBUG: 11:31:48.564: Device Virtual device with storage and identification for debugging scan type changed to 'swipe'                                                                                        
(fprintd:5769): fprintd-DEBUG: 11:31:48.564: Device Virtual device with storage and identification for debugging enroll stages changed to 6                           
(fprintd:5769): fprintd-DEBUG: 11:31:48.564: Finger present 0                                                                                
(fprintd:5769): fprintd-DEBUG: 11:31:48.564: Finger needed 0                                                                                 
(fprintd:5769): fprintd-DEBUG: 11:31:48.565: Got delay inhibitor for sleep.                                                                                           
(fprintd:5769): fprintd-DEBUG: 11:31:48.565: entering main loop                                                                                          
(fprintd:5769): fprintd-DEBUG: 11:31:48.565: D-Bus service launched with name: net.reactivated.Fprint                                                                 
(fprintd:5769): fprintd-DEBUG: 11:31:48.666: Requesting authorization from :1.0 to call method 'Claim' for device 'Virtual device with storage and identification for debugging'                                                                                      
(fprintd:5769): fprintd-DEBUG: 11:31:48.666: Getting authorization to perform Polkit action net.reactivated.fprint.device.verify                                                       
(fprintd:5769): fprintd-DEBUG: 11:31:48.667: Getting authorization to perform Polkit action net.reactivated.fprint.device.enroll                                      
(fprintd:5769): fprintd-DEBUG: 11:31:48.668: Authorization granted to :1.0 to call method 'Claim' for device 'Virtual device with storage and identification for debugging'!                                                         
(fprintd:5769): fprintd-DEBUG: 11:31:48.668: user 'cbuild' claiming the device: 0                                                                                                      
(fprintd:5769): libfprint-virtual_device-DEBUG: 11:31:48.668: 12922423295: ../libfprint/drivers/virtual-device.c:387                                                                                                                 
(fprintd:5769): libfprint-virtual_device_connection-DEBUG: 11:31:48.668: 12922423338: ../libfprint/drivers/virtual-device-listener.c:153                                                                                             
(fprintd:5769): libfprint-device-DEBUG: 11:31:48.668: Device reported open completion                                                                                                                                                
(fprintd:5769): libfprint-device-DEBUG: 11:31:48.668: Completing action FPI_DEVICE_ACTION_OPEN in idle!                                                                                
(fprintd:5769): libfprint-device-DEBUG: 11:31:48.669: Updated temperature model after 0.11 seconds, ratio 0.27 -> 0.27, active 0 -> 0, FP_TEMPERATURE_COLD -> FP_TEMPERATURE_COLD                                                    
(fprintd:5769): fprintd-DEBUG: 11:31:48.669: claimed device 0                                                                                                                                                                        
(fprintd:5769): fprintd-DEBUG: 11:31:48.670: Requesting authorization from :1.0 to call method 'VerifyStart' for device 'Virtual de
vice with storage and identification for debugging'                                                                                                                                                         
(fprintd:5769): fprintd-DEBUG: 11:31:48.670: Getting authorization to perform Polkit action net.reactivated.fprint.device.verify
                                                                                                                                                                                                                                     
(fprintd:5769): fprintd-WARNING **: 11:31:48.671: Authorization denied to :1.0 to call method 'VerifyStart' for device 'Virtual device with storage and identification for debugging': Not Authorized: net.reactivated.fprint.device.verify
(fprintd:5769): fprintd-DEBUG: 11:31:48.672: main loop completed                                                                                                                                                                     
(fprintd:5769): libfprint-virtual_device_storage-DEBUG: 11:31:48.673: 12922427888: ../libfprint/drivers/virtual-device-storage.c:253                                                                                                                                  
(fprintd:5769): libfprint-virtual_device-DEBUG: 11:31:48.673: 12922427910: ../libfprint/drivers/virtual-device.c:752

(fprintd:5769): libfprint-device-WARNING **: 11:31:48.673: User destroyed open device! Not cleaning up properly!
stderr:                                                          
test_unallowed_verify_with_enroll_claim (__main__.FPrintdVirtualDeviceStorageTest.test_unallowed_verify_with_enroll_claim) ... FAIL
/usr/lib/python3.12/unittest/case.py:597: ResourceWarning: unclosed file <_io.BufferedReader name=8>
  function(*args, **kwargs)                                      
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```

The return code of -6 apparently means the process exited with SIGABRT.

Most of libfprint's tests also error with SIGABRT, from my investigation possible because of a double-free somewhere, here's a sample backtrace:
```(lldb) bt
* thread #1, name = 'test-fp-device', stop reason = signal SIGABRT
  * frame #0: 0x000074ac5dc581de ld-musl-x86_64.so.1`abort [inlined] __syscall4(n=14, a1=1, a2=140736555305816, a3=0, a4=8) at syscall_arch.h:38:2
    frame #1: 0x000074ac5dc581d0 ld-musl-x86_64.so.1`abort at abort.c:23:2
    frame #2: 0x000074ac5dc631a6 ld-musl-x86_64.so.1`scudo::die() at linux.cpp:44:23
    frame #3: 0x000074ac5dc638c9 ld-musl-x86_64.so.1`scudo::ScopedErrorReport::~ScopedErrorReport(this=0x00007fffc862cbb8) at report.cpp:30:5
    frame #4: 0x000074ac5dc63c32 ld-musl-x86_64.so.1`scudo::reportInvalidChunkState(Action=Deallocating, Ptr=0x000074ac59a01390) at report.cpp:145:1
    frame #5: 0x000074ac5dc65ffb ld-musl-x86_64.so.1`scudo::Allocator<MuslConfig, &malloc_postinit()>::deallocate(this=<unavailable>, Ptr=<unavailable>, Origin=Malloc, DeleteSize=0, Alignment=<unavailable>) at combined.h:581:7
    frame #6: 0x000074ac5dc65e20 ld-musl-x86_64.so.1`__libc_free(ptr=<unavailable>) at wrappers.cpp:463:13 [artificial]
    frame #7: 0x000074ac5dc62820 ld-musl-x86_64.so.1`free(p=<unavailable>) at free.c:5:2 [artificial]
    frame #8: 0x000074ac5a2097d0 libglib-2.0.so.0`g_mutex_clear [inlined] g_mutex_impl_free(mutex=<unavailable>) at gthread-posix.c:124:3 [artificial]
    frame #9: 0x000074ac66140f8d libgusb.so.2`g_usb_context_dispose(object=0x000074ac597034c0) at gusb-context.c:143:2
    frame #10: 0x000074ac5dacbb12 libgobject-2.0.so.0`g_object_unref(_object=0x000074ac597034c0) at gobject.c:3894:7
    frame #11: 0x000074ac5a646926 libfprint-2.so.2`fp_context_finalize(object=0x000074ac5a400640) at fp-context.c:303:3
    frame #12: 0x000074ac5dacbbd1 libgobject-2.0.so.0`g_object_unref(_object=0x000074ac5a400640) at gobject.c:3941:11
    frame #13: 0x0000632c695d2847 test-fp-device`fpt_context_free(tctx=0x000074ac59e00dc0) at test-utils.c:145:3
    frame #14: 0x0000632c695d1d60 test-fp-device`glib_autoptr_clear_FptContext(_ptr=0x000074ac59e00dc0) at test-utils.h:44:1
    frame #15: 0x0000632c695d1bd8 test-fp-device`glib_autoptr_cleanup_FptContext(_ptr=0x00007fffc862cef0) at test-utils.h:44:1
    frame #16: 0x0000632c695d0905 test-fp-device`test_device_open_async at test-fp-device.c:45:1
    frame #17: 0x000074ac5a1c11f9 libglib-2.0.so.0`g_test_run_suite_internal at gtestutils.c:3161:15
    frame #18: 0x000074ac5a1c0c3e libglib-2.0.so.0`g_test_run_suite_internal(suite=0x000074ac59e00400, path=0x0000000000000000) at gtestutils.c:3256:16
    frame #19: 0x000074ac5a1c14ab libglib-2.0.so.0`g_test_run_suite_internal(suite=0x000074ac59e00670, path=0x0000000000000000) at gtestutils.c:0:3
    frame #20: 0x000074ac5a1c14ab libglib-2.0.so.0`g_test_run_suite_internal(suite=0x000074ac59e01150, path=0x0000000000000000) at gtestutils.c:0:3
    frame #21: 0x000074ac5a1bf132 libglib-2.0.so.0`g_test_run_suite(suite=0x000074ac59e01150) at gtestutils.c:3355:13
    frame #22: 0x000074ac5a1bee1a libglib-2.0.so.0`g_test_run at gtestutils.c:2462:7
    frame #23: 0x0000632c695d085c test-fp-device`main(argc=1, argv=0x00007fffc862d358) at test-fp-device.c:288:10
    frame #24: 0x000074ac5dc57b3a ld-musl-x86_64.so.1`libc_start_main_stage2(main=(test-fp-device`main at test-fp-device.c:267), argc=<unavailable>, argv=0x00007fffc862d358) at __libc_start_main.c:95:7
    frame #25: 0x0000632c695d0626 test-fp-device`_start + 22
```